src/include/homestore/chunk.h

```diff
@@ -13,10 +13,14 @@
  *
  *********************************************************************************/
 #pragma once
-#include "device/physical_dev.hpp"
+
+#include<homestore/homestore_decl.hpp>
+#include<homestore/blk.h>
+
 
 namespace homestore {
 class BlkAllocator;
+class PhysicalDev;
 
 class Chunk {
```

Comment:
I think we should expose a different structure that hides all of this; we don't need HomeObject to depend on the `sizeof(Chunk);` with all these dependencies if possible.

Can we hide this using a PIMPL so we can also change the implementation without worrying about the upper layers?

Reply:
yes, initially , i have considered using a pimpl and it is not graceful to expose all details and useless functions, but there are three interface might be used by upper layer

`available_blks()` :  this might be used by heapChunkSelector to sort the heap.

`user_private()` and `set_user_private()` : this might be used by upper layer to write and read private data, which make no sense to homestore.

another point is HeapChunkSelector should make sure all the shards of a PG should reside in the same pdev, so we might keep a heap for each of pdev, which might need `PhysicalDev*`

Reply:
Yes it had back and forth for several times.

The challenge is the caller of ChunkSelector need `chunk` to proceed , so we either need to return a `chunk` from Selector, or do some static_cast in the caller.
The former one expose chunk to external and as the `const chunk_info& info() const;` expose chunk_info then we need to expose `chunk_info` .
The latter one seems more worse IMHO.

Reply:
Why `available_blks()`, `user_private()` cannot be hidden in Pimpl ?

Reply:
we need to expose this to upper layer , so we can not hide in pimpl

Reply:
@szmyd @hkadayam  see if the current implementation with pimpl looks good.
`getInternalChunk` is used to expose Chunk for HS internal